### PR TITLE
UniV3 non-LP tokens support

### PR DESCRIPTION
**Pull request by julienmartinlevrai**

### CODE REVIEW

**Comment by julienmartinlevrai** [comment]
For the spécific case of wstETH, it should be converted to stETH before attempting to swap it. I am unsure whether this logic should be in a conditional inside this contract, if it should be in its own contract, or if it should be together with the [Curve contract](https://github.com/makerdao/exchange-callees/pull/18), which handles only this spécific collatéral. Thoughts ? @godsflaw @gbalabasquer @talbaneth

File: src/test/Simulation.t.sol [hunk 1]
```diff
@@ -21,6 +21,7 @@ import "ds-test/test.sol";
 import "dss-interfaces/Interfaces.sol";
 import { UniswapV2CalleeDai } from "../UniswapV2Callee.sol";
 import { UniswapV2LpTokenCalleeDai } from "../UniswapV2LpTokenCallee.sol";
+import { UniV3Callee } from "../UniV3Callee.sol";
```
Comment (brianmcmichael) [inline src/test/Simulation.t.sol:24]: There are a whole battery of tests for each of the other types: `UniswapV2Callee.t.sol`, `OasisDexCallee.t.sol`, are we missing a suite of tests at `UniV3Callee.t.sol`?

File: src/UniV3Callee.sol [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        uint256 amountIn;
+        uint256 amountOutMinimum;
+    }
+
+    function exactInput(UniV3Like.Params calldata params)
+        external payable returns (uint256 amountOut);
+}
+
+
+contract UniV3Callee {
```
Comment (brianmcmichael) [inline src/UniV3Callee.sol:56]: Kind of nitty, but this is `UniV3Callee` and the uniswap V2 one is called `UniswapV2Callee`.  Might want to be consistent on naming here.

File: src/test/Simulation.t.sol (hunk 1 above)
Comment (brianmcmichael) [reply src/test/Simulation.t.sol:24]: Also `UniV3CalleeCharter.t.sol`

**Review by brianmcmichael — CHANGES_REQUESTED** [review]
Some minor changes requested. Also tests that mirror the existing callee options.

File: src/test/Simulation.t.sol (hunk 1 above)
Comment (julienmartinlevrai) [reply src/test/Simulation.t.sol:24]: `UniswapV2Callee.t.sol` and `OasisDexCallee.t.sol` hold legacy tests with no mainnet fork simulation and instead use a mockup exchange function. Hence the existence of `Simulation.t.sol`, which performs simulation tests for UniswapV2. In this PR I added two tests for V3: `testTakeLinkV3Profit` and `testTakeLinkV3NoProfit`. Please let me know if you consider there are things we should be testing which we are currently not.

File: src/UniV3Callee.sol [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    function approve(address, uint256) external;
+    function transfer(address, uint256) external;
+    function balanceOf(address) external view returns (uint256);
+}
+
+interface CharterManagerLike {
+    function exit(address crop, address usr, uint256 val) external;
+}
+
+interface UniV3Like {
```
[the remark this answers is not in the document]
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol]: I changed it to `UniV3RouterLike`

File: src/UniV3Callee.sol [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+
+contract UniV3Callee {
+    UniV3Like               public uniV3;
+    DaiJoinLike             public daiJoin;
+    TokenLike               public dai;
+
+    uint256                 public constant RAY = 10 ** 27;
+
+    function add(uint x, uint y) internal pure returns (uint z) {
```
[the remark this answers is not in the document]
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol]: done

File: src/UniV3Callee.sol [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    function balanceOf(address) external view returns (uint256);
+}
+
+interface CharterManagerLike {
+    function exit(address crop, address usr, uint256 val) external;
+}
+
+interface UniV3Like {
+    
+    struct Params {
```
[the remark this answers is not in the document]
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol]: done

File: src/UniV3Callee.sol [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        require((z = x + y) >= x, "ds-math-add-overflow");
+    }
+    function sub(uint x, uint y) internal pure returns (uint z) {
+        require((z = x - y) <= x, "ds-math-sub-underflow");
+    }
+    function divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = add(x, sub(y, 1)) / y;
+    }
+
+    constructor(address uniV3Addr_, address daiJoin_) public {
```
[the remark this answers is not in the document]
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol]: done

File: src/UniV3Callee.sol (hunk 2 above)
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol:56]: holding off on this one for the very end as I don’t know how GitHub handles review comments in renamed files.

File: src/UniV3Callee.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (talbaneth) [reply src/UniV3Callee.sol]: Please apply such changes also to the curve callee contract.

File: src/UniV3Callee.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (talbaneth) [reply src/UniV3Callee.sol]: Also here - apply to curve calleee please.

**Comment by talbaneth** [comment]
@julienmartinlevrai I reviewed it and aside of @brianmcmichael's comments it looks good.
Suggest to merge the curve work at this point.
Note the comment I made about applying the renamings (like RouterLike) also to the curve work.

**Comment by julienmartinlevrai** [comment]
@talbaneth thank you for your review. I will not merge the Curve PR here, rather I will merge this and then rebase the Curve PR to master.

File: src/UniV3Callee.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol]: done

File: src/UniV3Callee.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol]: done

**Comment by julienmartinlevrai** [comment]
@brianmcmichael @talbaneth PR ready for your approvals

**Review by talbaneth — APPROVED** [review]
LGTM

File: src/UniV3Callee.sol (hunk 2 above)
Comment (julienmartinlevrai) [reply src/UniV3Callee.sol]: done

**Review by godsflaw — APPROVED** [review]
LGTM integrating into the `auction-demo-keeper` now

**Comment by julienmartinlevrai** [comment]
@brianmcmichael @godsflaw @talbaneth thank you for your reviews, merging now

**Pull request merged**